### PR TITLE
Introduce new BM25Vectorizer

### PR DESCRIPTION
**Pull request by x65han**
- Introduce new BM25Vectorizer
- Refactor Common logic of vectorizer to `Class VectorizerBase`
- Add test case in `20newsgroups-replication.py` with f1 score assertion
- Replace `MultinomialNB` with `LogisticRegression`, which renders better results

----------------------------

tfidf vectorizer with Linear Regression model score: 0.8359058
bm25 vectorizer with Linear Regression model score: 0.8421606

----------------------------
```
python scripts/20newsgroups-replication.py --vectorizer TfidfVectorizer
python scripts/20newsgroups-replication.py --vectorizer BM25Vectorizer
```

### CODE REVIEW

File: pyserini/vectorizer.py [hunk 1]
```diff
@@ -22,8 +22,8 @@
 from pyserini.index import pyutils
 
 
-class TfidfVectorizer:
-    """Wrapper class for tf-idf vectorizer implemented on top of Pyserini.
+class VectorizerBase:
```
Comment (lintool) [inline pyserini/vectorizer.py:25]: Just `Vectorizer`? no base?

File: pyserini/vectorizer.py [hunk 2]
```diff
@@ -36,23 +36,21 @@ class TfidfVectorizer:
     """
 
     def __init__(self, lucene_index_path: str, min_df: int = 1, verbose: bool = False) -> None:
```
Comment (lintool) [inline pyserini/vectorizer.py:38]: I suppose `min_df` is still applicable for filtering the vocab, even if we're ultimately computing BM25. Is that the intention?

File: pyserini/vectorizer.py [hunk 3]
```diff
@@ -103,4 +118,53 @@ def get_vectors(self, doc_ids: List[str]):
...
+    def __init__(self, lucene_index_path: str, min_df: int = 1, verbose: bool = False) -> None:
+        super().__init__(lucene_index_path, verbose, min_df)
+
+    def get_vectors(self, doc_ids: List[str]):
+        """Get the bm25 vectors given a list of doc_ids
+
+        Parameters
+        ----------
+        doc_ids : List[str]
+            The piece of text to analyze.
```
Comment (lintool) [inline pyserini/vectorizer.py:146]: isn't this the list of docids?
btw, can you change `doc_ids` to `docids`? I use the latter consistently.

File: pyserini/vectorizer.py [hunk 4]
```diff
@@ -103,4 +118,53 @@ def get_vectors(self, doc_ids: List[str]):
...
+
+        Returns
+        -------
+        csr_matrix
+            L2 normalized sparse matrix representation of bm25 vectors
```
Comment (lintool) [inline pyserini/vectorizer.py:151]: capitalize BM25 in this context.

File: pyserini/vectorizer.py (hunk 2 above)
Comment (x65han) [reply pyserini/vectorizer.py:38]: Yes, with BM25Vectorizer we can still use `min_df` to filter out rare words when building our vocabulary.

File: scripts/20newsgroups-replication.py [hunk 5]
```diff
@@ -19,43 +21,44 @@ def get_info(path):
...
-        'sci.space',
-        'soc.religion.christian',
-        'talk.politics.guns',
-        'talk.politics.mideast',
-        'talk.politics.misc',
-        'talk.religion.misc',
-    ]
+    parser = argparse.ArgumentParser(
+        description='Replication script of pyserini vectorizer')
+    parser.add_argument('-vectorizer', type=VectorizerType, required=True, help='which vectorizer to use')
```
Comment (lintool) [inline scripts/20newsgroups-replication.py:49]: `--vectorizer`? per what we agreed to? https://github.com/lintool/guide/blob/master/coding-style.md

File: scripts/20newsgroups-replication.py [hunk 6]
```diff
@@ -19,43 +21,44 @@ def get_info(path):
...
     target_to_index = {t: i for i, t in enumerate(target_names)}
 
     train_docs, train_labels = get_info('./20newsgroups/20news-bydate-train/')
     test_docs, test_labels = get_info('./20newsgroups/20news-bydate-test/')
 
-    vectorizer = TfidfVectorizer(
-        './20newsgroups/lucene-index.20newsgroup.pos+docvectors+raw', min_df=5)
+    lucene_index_path = '20newsgroups/lucene-index.20newsgroup.pos+docvectors+raw'
+    vectorizer = None
+    if args.vectorizer == VectorizerType.TFIDF:
```
Comment (lintool) [inline scripts/20newsgroups-replication.py:62]: how about we just pass in the class name from the cmd line and init from string? e.g., https://stackoverflow.com/questions/4821104/dynamic-instantiation-from-string-name-of-a-class-in-dynamically-imported-module

that way we don't need this enum...

Comment (x65han) [reply scripts/20newsgroups-replication.py:62]: Interesting. I didn't know this is possible.

**Pull request merged**